File: J9Ch/src/S_6_Dynamic_Problem_II/Optional_11/python_cpp/_440_Backpack_III_Hard.py

class Solution:
    # @param {int[]} A an integer array
    # @param {int[]} V an integer array
    # @param {int} m an integer
    # @return {int} an array
    def backPackIII(self, A, V, m):
        # Write your code here
        f = [0 for i in xrange(m+1)]
        for (a, v) in zip(A, V):
            for j in xrange(a, m+1):
                if f[j - a] + v > f[j]:
                    f[j] = f[j - a] + v
        return f[m]

File: J9Ch/src/S_6_Dynamic_Problem_II/Optional_11/python_cpp/_440_Backpack_III_Hard.cpp

class Solution {
public:
    /**
     * @param A an integer array
     * @param V an integer array
     * @param m an integer
     * @return an array
     */
    int backPackIII(vector<int>& A, vector<int>& V, int m) {
        // Write your code here
        int n = A.size();
        int* f = new int[m+1];
        for (int i = 0; i <= m; ++i) f[i] = 0;
        for (int i = 0; i < n; ++i)
            for (int j = A[i]; j <=m; ++j)
                if (f[j - A[i]] + V[i] > f[j])
                    f[j] = f[j - A[i]] + V[i];

        return f[m];
    }
};